$\text{IMPORTS}$

In [1]:
import cvxpy as cp
import numpy as np

$\textbf{QUESTION 1}$
$$

$$

In [2]:
A1 = np.array([40, 0, 0])
A2 = np.array([0, 16, 0])
A3 = np.array([0, 0, 12])
c_b = np.array([1, 1, 1])

A = np.matrix([A1, A2, A3]).T
B = np.matrix([A1, A2, A3]).T
inv_b = np.linalg.inv(B)
y_hat = np.array(c_b @ inv_b)
print(f"\nBasis Matrix:\n{B=}\nInverse of Basis Matrix:\n{inv_b=}\n{y_hat=}")
print("\nOptimal Solution:\nx1 = 3.75\nx2 = 6.25\nx3 = 6.67")
print("\nOptimal Value:\n16.67")


Basis Matrix:
B=matrix([[40,  0,  0],
        [ 0, 16,  0],
        [ 0,  0, 12]])
Inverse of Basis Matrix:
inv_b=matrix([[0.025     , 0.        , 0.        ],
        [0.        , 0.0625    , 0.        ],
        [0.        , 0.        , 0.08333333]])
y_hat=array([[0.025     , 0.0625    , 0.08333333]])

Optimal Solution:
x1 = 3.75
x2 = 6.25
x3 = 6.67

Optimal Value:
16.67


$\textbf{QUESTION 2}\\
\text{Iteration 1}$
$$
\text{Solve RMP:} \\
min \ \{x_1 + x_2 + x_3\} \\
S.T.\\
\begin{bmatrix}
40\\
0\\
0
\end{bmatrix}x_1 +
\begin{bmatrix}
 0\\
 16\\
 0
\end{bmatrix}x_2 +
 \begin{bmatrix}
0\\
0\\
12
\end{bmatrix}x_3 =
\begin{bmatrix}
150\\
100\\
80
\end{bmatrix} \\
x_1, x_2, x_3 \ge 0
$$

In [3]:
x1 = cp.Variable(1)
x2 = cp.Variable(1)
x3 = cp.Variable(1)

obj = cp.Minimize(x1 + x2 + x3)
constraints = [40 * x1 == 150,
               16 * x2 == 100, 
               12 * x3 == 80,
               x1 >=0,
               x2 >=0,
               x3 >=0]

roll = cp.Problem(obj, constraints)
roll.solve()
print(f"The Optimal Value:\n{roll.value}")

x = np.array([v[0].value for i, v in enumerate([x1, x2, x3])])
print(f"\nOptimal Solution:\n{x=}")

cvx_yhat = np.array([-constraints[i].dual_value[0] for i, v in enumerate(constraints)])
print(f"\nOptimal Dual Solution:\n{cvx_yhat=}")

print(f"\nBasis Matrix:\n{B=}\nInverse of Basis Matrix:\n{inv_b=}")

The Optimal Value:
16.66666666666666

Optimal Solution:
x=array([3.75      , 6.25      , 6.66666667])

Optimal Dual Solution:
cvx_yhat=array([ 2.50000000e-02,  6.25000000e-02,  8.33333333e-02, -6.34729978e-10,
       -4.44359035e-10, -4.00182754e-10])

Basis Matrix:
B=matrix([[40,  0,  0],
        [ 0, 16,  0],
        [ 0,  0, 12]])
Inverse of Basis Matrix:
inv_b=matrix([[0.025     , 0.        , 0.        ],
        [0.        , 0.0625    , 0.        ],
        [0.        , 0.        , 0.08333333]])


$\textbf{QUESTION 3}$

$$
\text{Write the Knapsack Problem:}\\
z = max \{0.025a_1 + 0.0625a_2 + 0.0833a_3 \}\\
S.T.\\
5a_1 + 12a_2 + 16a_3 \le 200\\
a_1, a_2, a_3 \ge 0
$$

$\textbf{QUESTION 4}$

$$
\text{Solve the Knapsack Problem:}
$$

In [4]:
a1 = cp.Variable(1, integer=True)
a2 = cp.Variable(1, integer=True)
a3 = cp.Variable(1, integer=True)

expr = (0.025 * a1) + (0.0625 * a2) + (0.0833 * a3)

next_obj = cp.Maximize(expr)
next_constraints = [(5 * a1) + (12 * a2) + (16 * a3) <= 200,
                    a1 >= 0,
                    a2 >= 0,
                    a3 >= 0]

next_roll = cp.Problem(next_obj, next_constraints)
next_roll.solve()
print(f"The Optimal Value:\n{next_roll.value}")

A4 = np.matrix([v[0].value for i, v in enumerate([a1, a2, a3])]).T
print(f"\nOptimal Solution:\n{A4=}")

new_A = np.append(A, A4, axis=1)
new_B = np.append(B, A4, axis=1)

print(f"New pattern:\n{new_A=}")



The Optimal Value:
1.0416

Optimal Solution:
A4=matrix([[ 0.],
        [14.],
        [ 2.]])
New pattern:
new_A=matrix([[40.,  0.,  0.,  0.],
        [ 0., 16.,  0., 14.],
        [ 0.,  0., 12.,  2.]])


$$
\text{The optimal solution (z):}\ 1.0416 \Rightarrow \\
1-z \lt 0 \Rightarrow \\
\text{Column generation algorithm does not terminate.} \Rightarrow \\
\text{We say the new column to enter:} \ A_4 = [0, 14, 2] \Rightarrow \\~\\
\text{New Pattern (A):}
\begin{bmatrix}
40 & 0 & 0 & 0\\
0 & 16 & 0 & 14\\
0 & 0 & 12 & 2
\end{bmatrix}
$$

$\textbf{QUESTION 5}\\
\text{Iteration 2}\\
$

$$
\text{Solve New RMP:} \\
min \ \{x_1 + x_2 + x_3 + x_4 \} \\
S.T.\\
\begin{bmatrix}
40\\
0\\
0
\end{bmatrix}x_1 +
\begin{bmatrix}
 0\\
 16\\
 0
\end{bmatrix}x_2 +
 \begin{bmatrix}
0\\
0\\
12
\end{bmatrix}x_3 +
\begin{bmatrix}
0\\
14\\
2
\end{bmatrix}x_4 =
\begin{bmatrix}
150\\
100\\
80
\end{bmatrix} \\
x_1, x_2, x_3, x_4 \ge 0
$$

In [5]:
x1 = cp.Variable(1)
x2 = cp.Variable(1)
x3 = cp.Variable(1)
x4 = cp.Variable(1)

obj = cp.Minimize(x1 + x2 + x3 + x4)
constraints = [(40 * x1) == 150,
               (16 * x2) + (14 * x4) == 100, 
               (12 * x3) + (2 * x4) == 80,
               x1 >=0,
               x2 >=0,
               x3 >=0,
               x4 >=0]

new_roll = cp.Problem(obj, constraints)
new_roll.solve()
print(f"The Optimal Value:\n{new_roll.value}")

x = np.array([v[0].value for i, v in enumerate([x1, x2, x3, x4])])
print(f"\nOptimal Solution:\n{x.round(4)=}")

new_B = np.delete(new_B, obj=1,  axis=1)
new_c_b = np.array([1, 1, 1])
new_inv_b = np.linalg.inv(new_B)
new_yhat = np.array(new_c_b @ new_inv_b)

print(f"\nBasis Matrix:\n{new_B=}\n\nInverse of Basis Matrix:\n{new_inv_b=}\
    \n\nOptimal Dual Solution:\n{new_yhat=}")

The Optimal Value:
16.369047662144975

Optimal Solution:
x.round(4)=array([3.75  , 0.    , 5.4762, 7.1429])

Basis Matrix:
new_B=matrix([[40.,  0.,  0.],
        [ 0.,  0., 14.],
        [ 0., 12.,  2.]])

Inverse of Basis Matrix:
new_inv_b=matrix([[ 0.025     ,  0.        ,  0.        ],
        [ 0.        , -0.01190476,  0.08333333],
        [ 0.        ,  0.07142857,  0.        ]])    

Optimal Dual Solution:
new_yhat=array([[0.025     , 0.05952381, 0.08333333]])


$$\text{Solve New Knapsack Problem:}
$$

In [6]:
a1 = cp.Variable(1, integer=True)
a2 = cp.Variable(1, integer=True)
a3 = cp.Variable(1, integer=True)

expr = (0.025 * a1) + (0.0595 * a2) + (0.0833 * a3)

next_obj = cp.Maximize(expr)
next_constraints = [(5 * a1) + (12 * a2) + (16 * a3) <= 200,
                    a1 >= 0,
                    a2 >= 0,
                    a3 >= 0]

next_roll = cp.Problem(next_obj, next_constraints)
next_roll.solve()
print(f"The Optimal Value:\n{next_roll.value}")

A5 = np.matrix([v[0].value for i, v in enumerate([a1, a2, a3])]).T
print(f"\nOptimal Solution:\n{A5=}")

The Optimal Value:
1.0352999999999999

Optimal Solution:
A5=matrix([[ 0.],
        [ 2.],
        [11.]])


$$
\text{The optimal solution (z):}\ 1.03562\Rightarrow \\
1-z \lt 0 \Rightarrow \\
\text{Column generation algorithm does not terminate.} \Rightarrow \\
\text{We say the new column to enter:} \ A_5 = [0, 2, 11] \Rightarrow \\~\\
$$

In [7]:
# APPEND TO A AND B
new_A = np.append(new_A, A5, axis=1)
new_B = np.append(new_B, A5, axis=1)

In [8]:
print(f"New pattern:\n{new_A}")

New pattern:
[[40.  0.  0.  0.  0.]
 [ 0. 16.  0. 14.  2.]
 [ 0.  0. 12.  2. 11.]]


$
\text{Iteration 3}
$

$$
\text{Solve New RMP:} \\
min \ \{x_1 + x_2 + x_3 + x_4 + x_5 \} \\
S.T.\\
\begin{bmatrix}
40\\
0\\
0
\end{bmatrix}x_1 +
\begin{bmatrix}
 0\\
 16\\
 0
\end{bmatrix}x_2 +
 \begin{bmatrix}
0\\
0\\
12
\end{bmatrix}x_3 +
\begin{bmatrix}
0\\
14\\
2
\end{bmatrix}x_4
+
\begin{bmatrix}
0\\
2\\
11
\end{bmatrix}x_5
=
\begin{bmatrix}
150\\
100\\
80
\end{bmatrix} \\
x_1, x_2, x_3, x_4, x_5 \ge 0
$$

In [9]:
# NEW RMP
x1 = cp.Variable(1)
x2 = cp.Variable(1)
x3 = cp.Variable(1)
x4 = cp.Variable(1)
x5 = cp.Variable(1)

obj = cp.Minimize(x1 + x2 + x3 + x4 + x5)
constraints = [(40 * x1) == 150,
               (16 * x2) + (14 * x4) + (2 * x5) == 100, 
               (12 * x3) + (2 * x4) + (11 * x5) == 80,
               x1 >=0,
               x2 >=0,
               x3 >=0,
               x4 >=0,
               x5 >=0]

new_roll = cp.Problem(obj, constraints)
new_roll.solve()
print(f"The Optimal Value:\n{new_roll.value}")

x = np.array([v[0].value for i, v in enumerate([x1, x2, x3, x4, x5])])
print(f"\nOptimal Solution:\n{x.round(4)=}")

# UPDATE B
new_B = np.delete(new_B, obj=[1],  axis=1)

new_c_b = np.array([1, 1, 1])
new_inv_b = np.linalg.inv(new_B)
new_yhat = np.array(new_c_b @ new_inv_b)

print(f"\nNew Basis Matrix:\n{new_B=}\n\nInverse of Basis Matrix:\n{new_inv_b=}\
     \n\nOptimal Dual Solution:\n{new_yhat=}")

The Optimal Value:
16.15000000103377

Optimal Solution:
x.round(4)=array([3.75  , 0.    , 0.    , 6.2667, 6.1333])

New Basis Matrix:
new_B=matrix([[40.,  0.,  0.],
        [ 0., 14.,  2.],
        [ 0.,  2., 11.]])

Inverse of Basis Matrix:
new_inv_b=matrix([[ 0.025     ,  0.        ,  0.        ],
        [ 0.        ,  0.07333333, -0.01333333],
        [ 0.        , -0.01333333,  0.09333333]])     

Optimal Dual Solution:
new_yhat=array([[0.025, 0.06 , 0.08 ]])


$$\text{Solve New Knapsack Problem:}
$$

In [10]:
a1 = cp.Variable(1, integer=True)
a2 = cp.Variable(1, integer=True)
a3 = cp.Variable(1, integer=True)

expr = (0.025 * a1) + (0.06 * a2) + (0.08 * a3)

next_obj = cp.Maximize(expr)
next_constraints = [(5 * a1) + (12 * a2) + (16 * a3) <= 200,
                    a1 >= 0,
                    a2 >= 0,
                    a3 >= 0]

next_roll = cp.Problem(next_obj, next_constraints)
next_roll.solve()
print(f"The Optimal Value:\n{next_roll.value}")

A6 = np.matrix([v[0].value for i, v in enumerate([a1, a2, a3])]).T
print(f"\nOptimal Solution:\n{A6=}")

The Optimal Value:
1.0

Optimal Solution:
A6=matrix([[40.],
        [ 0.],
        [ 0.]])


$$
\text{The optimal solution (z):}\ 1.0\Rightarrow \\
1-1 = 0 \Rightarrow \\
\text{Column generation algorithm terminates.}
$$

$\textbf{QUESTION 6}$

$$
\text{Final Optimal Solution of Knapsack: } A_6 =
\begin{bmatrix}
a_1 & a_2 & a_3
\end{bmatrix} \Rightarrow
\begin{bmatrix}
40 & 0 & 0
\end{bmatrix} \\~\\
\text{Final Optimal Solution of X: }
\begin{bmatrix}
x_1 & x_2 & x_3 & x_4 & x_5
\end{bmatrix} \Rightarrow
\begin{bmatrix}
3.75 & 0 & 0 & 6.2667 & 6.1333
\end{bmatrix}
\\~\\
\text{Final Optimal Basis: }
\begin{bmatrix}
40 & 0 & 0 \\
0 & 14 &  2 \\
0 &  2 & 11 \\
\end{bmatrix} 
\\~\\
\text{Final Optimal Objective Value: }1.0
$$